### ISSUE TITLE
In superresolution mode, backgrounds tend to grey despite the source background being either white or black

### ISSUE THREAD

**Issue by ramayer** [ramayer/Elucidated-Diffusion-Model-vibe-coded#6]
User:

This bugs me..... in both SR mode and unconditioned, almost all my backgrounds tend toward a middling grey.    Most of my source training images have either black or white background.   It's weird to see the super resolution path take an image with a totally black background and output one with reasonably correct structure, and reasonably correct hair color, but a light grey background.   When it does this, it lightens the figure too.     I'd have thought a huge color shift on a background should have triggered a huge loss.

Claude

This looks like a mean regression problem in the diffusion model — the network is collapsing toward the average of the training data distribution rather than respecting the conditioning signals, which suggests an issue with the EDM preconditioning or sigma_data calibration that's causing the model to ignore the black/white background distinctions it clearly learned. The mode-averaging problem gets worse with deterministic samplers that lack stochasticity—they can get trapped near the mean at high sigma steps, especially with few sampling steps. Even when the thumbnail clearly shows black background pixels, if the injection strength isn't fully overriding the model's prediction in those background regions (where the signal is weaker), the model defaults to averaging the bimodal distribution and outputs grey instead of committing to either black or white. GroupNorm is likely the culprit here. Since it normalizes across spatial dimensions and channel groups together, a large flat background region could interact strangely with the normalization — the uniformity of the background might get suppressed or redistributed in a way that masks the color information, even though the gradient signal from the mismatch should theoretically be huge. ift parameters are learned per-channel and fixed across the image, they can't adaptively encode whether the background is globally black or white—that information gets lost when GroupNorm strips the mean. So the network has to rely on AdaLN's time-conditioned affine parameters to recover that global brightness signal, but those are also applied uniformly across space, meaning the model needs to encode "black background" vs "white background" through the conditioning mechanism rather than through spatially-varying activations. So the real issue is that GroupNorm strips away the global brightness signal at every layer—it's literally designed to remove the per-sample mean that would distinguish a black background from a white one. …

Good question, and I think your intuition about "huge loss" is actually backwards in a way that explains a lot here — plus this connects directly to something we were about to change anyway.

Why grey doesn't trigger a huge loss — it minimizes it.

If your training set has a genuine bimodal split (some images black background, some white), and at a given point in denoising the network has no way yet to tell which mode this particular sample belongs to, then predicting grey is the mathematically optimal thing to output — not a compromise that costs you, but literally the minimum-expected-squared-error answer under that uncertainty. Squared error penalizes confident wrong answers far more than it penalizes a hedge: confidently predicting black when the truth is white costs you (black - white)² — a large number — while predicting the mean costs you a smaller, consistent penalty every time, regardless of which mode was true. Averaged over the whole bimodal training distribution, the mean has lower expected loss than either confident guess. So "the network is outputting a suspicious grey" isn't evidence something's broken — under genuine ambiguity, grey is what a correctly-trained MMSE estimator is supposed to output. The question isn't "why doesn't a big error get punished," it's "why is the ambiguity still present at a point where it shouldn't be."

For the unconditioned path, some of this is just expected diffusion behavior at high sigma (this is the standard "regression to the mean" phenomenon — at very high noise, the optimal denoiser output for an ambiguous region really is the population average), and the real question is whether it resolves as sigma decreases toward the end of sampling, or gets stuck. That's a separate, known issue (sampler step count / stochasticity), not something the training loss should need to fix directly.

For the SR path, though, this is more concerning — and I think it's a direct consequence of the additive injection we're still using, not a training-time issue. The thumbnail should resolve this ambiguity completely, from the very first denoising step: background color is exactly the kind of information that survives downsampling to 64×64 perfectly (it's the lowest possible spatial frequency there is). So if the SR output is still averaging toward grey, the thumbnail's clear signal isn't actually overriding the backbone's own uncertainty-driven pull toward the mean — because right now, injection is still h + gate * content. That's additive, not substitutive: even a large, confident, correct signal from the thumbnail only gets added to whatever h already wants to say, and if h's own high-sigma estimate is pulling toward grey (its learned default under ambiguity), the injected content doesn't replace that pull, it just offsets it by some amount. If the gate hasn't yet learned a large enough magnitude to fully cancel that default, what you'd see is exactly a partial blend — grey nudged partway toward black — which matches what you're describing.

This is precisely the failure mode the gated-interpolation switch we discussed last turn was meant to fix: output = (1 - gate) * h + gate * content lets gate → 1 fully replace h's contribution rather than merely adding against it. With additive injection, there's no value of gate that fully removes h's own vote; with interpolation, there is. I'd treat what you're seeing as good evidence for prioritizing that switch, not just a nice-to-have.

On why the whole figure lightens, not just the background — this part I think is architectural, and worth knowing about even after the interpolation fix: GroupNorm computes its mean and variance across the entire spatial extent of a group, jointly with the channels in that group — not per-channel, not per-pixel. If background pixels make up a large fraction of the image and their value drifts (toward grey, in this case), that shifts the group's overall computed mean, which then affects how every other pixel in that group — including the figure — gets normalized and rescaled. In other words, GroupNorm structurally couples background brightness to foreground brightness within a block, because both are being normalized against one shared, spatially-pooled statistic. That's a plausible, concrete reason a grey-shifted background would drag the figure's brightness along with it, independent of whatever's causing the grey in the first place.

I'd fix the injection mechanism first (that's already the planned next-restart change) and see how much of this resolves on its own — if the figure-lightening persists even once backgrounds are correctly saturated, that'd be a good signal the GroupNorm coupling is contributing meaningfully on its own and worth addressing separately (e.g. switching flat/background-heavy blocks to a norm that doesn't pool over the full spatial extent).

<img width="64" height="64" alt="Image" src="https://github.com/user-attachments/assets/541975a8-6ce8-46da-979d-7b5a82645ca9" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/74b1c6d5-5b73-47fc-a59e-e8fc837be13b" />

**Comment by ramayer** [comment]
Another example:

<img width="64" height="64" alt="Image" src="https://github.com/user-attachments/assets/870f7bc6-b1a0-4255-ad73-7b0bee519492" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/9b53ba54-b0ee-4d5f-86e2-1d796ccc0b8e" />
